libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+        };
+
+        /// <summary>
+        /// The Azure Storage Blob Container where entities will be stored
+        /// </summary>
+        public Lazy<CloudBlobContainer> Container { get; private set; }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="dataConnectionString">Azure Storage connection string</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(string dataConnectionString, string containerName)
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+        }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="storageAccount">Azure CloudStorageAccount instance</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null) throw new ArgumentNullException(nameof(storageAccount));
+
+            // Checks if a container name is valid
+            NameValidator.ValidateContainerName(containerName);
+
+            this.Container = new Lazy<CloudBlobContainer>(() =>
+              {
+                  var blobClient = storageAccount.CreateCloudBlobClient();
+                  var container = blobClient.GetContainerReference(containerName);
+                  container.CreateIfNotExistsAsync().Wait();
+                  return container;
+              });
+        }
+
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
+            {
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
+                await blobReference.DeleteAsync(
+                    DeleteSnapshotsOption.None, new AccessCondition { IfMatchETag = "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);
+            }

```

Comment:
I would think we want to delete all the blobs in parallel. So instead of `await`ing the call in the loop, add the `Task` to a `var deleteTasks = new List<Task>(keys.Length);` and then `await  Task.WhenAll(deleteTasks)` right after the loop.

Reply:
I've commented on that a few times. Doing the loads / deletes in parallel seems like it would make a real difference. 

Juan?